File: MockInterview.Frontend/src/types/userDto.ts
import {ContactDto} from "@/types/contactDto";
import {TalentDto} from "@/types/talentDto";
import {IMappable, Util} from "@/types/common";

export interface IUserDto {
    id: number;
    firstName: string;
    lastName: string;
    emailAddress: string;
    gender: string;
    dateOfBirth: Date;
    photoStorageId: number;
    password: string;
    contact: ContactDto;
    talent: TalentDto;
}

export class UserDto implements IMappable<IUserDto> {
    id!: number;
    firstName!: string;
    lastName!: string;
    emailAddress!: string;
    gender!: string;
    dateOfBirth!: Date;
    photoStorageId!: number;
    password!: string;
    contact!: ContactDto | null;
    talent!: TalentDto | null;

    mapFrom(data: IUserDto) {
        this.id = data.id;
        this.firstName = data.firstName;
        this.lastName = data.lastName;
        this.emailAddress = data.emailAddress;
        this.gender = data.gender;
        this.dateOfBirth = data.dateOfBirth;
        this.photoStorageId = data.photoStorageId;
        this.password = data.password;
        this.contact = Util.mapObject(data.contact, ContactDto);
        this.talent = Util.mapObject(data.talent, TalentDto);
    }
}

File: MockInterview.Backend/MockInterview.BLL/Models/DTOs/UserDto.cs
using MockInterview.BLL.Models.Common;

namespace MockInterview.BLL.Models.DTOs;

/// <summary>
/// Represents system user
/// </summary>
public class UserDto : IEntityDto
{
    public long Id { get; set; }

    /// <summary>
    /// First name
    /// </summary>
    public string FirstName { get; set; } = null!;

    /// <summary>
    /// Lastname 
    /// </summary>
    public string LastName { get; set; } = null!;

    /// <summary>
    /// Email address
    /// </summary>
    public string EmailAddress { get; set; } = null!;

    /// <summary>
    /// Gender
    /// </summary>
    public string Gender { get; set; } = null!;

    /// <summary>
    /// Date of birth
    /// </summary>
    public DateTime DateOfBirth { get; set; }
    
    /// <summary>
    /// Stored photo Id
    /// </summary>
    public long PhotoStorageId { get; set; }

    /// <summary>
    /// Account password
    /// </summary>
    public string Password { get; set; } = null!;
    
    /// <summary>
    /// Contact projection
    /// </summary>
    public ContactDto? Contact { get; set; }
    
    /// <summary>
    /// Talent projection
    /// </summary>
    public TalentDto? Talent { get; set; }

    /// <summary>
    /// User Role projection
    /// </summary>
    public UserRoleDto? Role { get; set;} 
}